### ISSUE TITLE
Integration: Contact Form 7

### ISSUE THREAD

**Issue by michaelbeil** [AffiliateWP/AffiliateWP#1336]
We get requests for better integration with Contact Form 7.

Let's look into this and see if we can make it happen. 

Current partial integration: http://docs.affiliatewp.com/article/657-contact-form-7.

**Comment by sumobi** [comment]
This is going to be a bit tricky since CF7 doesn't have a "total" field.

It also doesn't store form entries by default so we won't be able to add the reference unless we also integrate with their other official plugin: https://wordpress.org/plugins/flamingo/

There's no official payment gateways so we'll need to support a 3rd party plugin such as: https://wordpress.org/plugins/contact-form-7-paypal-add-on/ or https://wordpress.org/plugins/contact-form-7-paypal-extension/

**Comment by pippinsplugins** [comment]
It would be pretty cool to support Flamingo and the payment add-ons.

**Comment by sumobi** [comment]
Yeah, we could definitely leave the reference field empty and include in our docs that Flamingo is required to have reference and link to entry.

**Comment by pippinsplugins** [comment]
Another request for it: https://secure.helpscout.net/conversation/245908794/41278/?folderId=205524

**Comment by pippinsplugins** [comment]
Another request https://secure.helpscout.net/conversation/282208045/47589/?folderId=205524

**Comment by ramiabraham** [comment]
Can't be worse than...

I'll take it :)

**Comment by ramiabraham** [comment]
Not a _request_, but the current iteration is certainly causing some confusion with customers, as (at least this customer) understandably thinks it's a _full_ integration:

https://secure.helpscout.net/conversation/283660919/47966/?folderId=297160

**Comment by ramiabraham** [comment]
@AffiliateWP/core General CF7 outline

### Current state of CF7 

CF7 has come a long way since last I've worked with it. There were hardly any hooks, and most of it was procedurally written, offering little extendability. Today, it's quite different:

- 24 actions
- 53 filters
- A standardized object structure, loosely mvc.
- All submissions are an instanceof `WPCF7_Submission`
- Using the _Flamingo_ plugin as an example, now there's a filter to determine the total, it is: `wpcf7_flamingo_get_value`.
- General modification of submission data will likely run through `wpcf7_posted_data`.

---

### Roadmap

- [ ] Define a new `WPCF7_Integration` base integration via our standard integration method.
- [ ] Via the CF7 integration as well as standalone hooks such as `wpcf7_contact_form_properties`, define functionality consistent with the AFNF for Ninja Forms 2.x integration (inline meta boxes and options).
- [ ] Test general integration functionality
- [ ] Create cross-integrations for the following CF7 add-ons:
    - https://wordpress.org/plugins/flamingo/
    - https://wordpress.org/plugins/contact-form-7-paypal-add-on/ `<or>`
    - https://wordpress.org/plugins/contact-form-7-paypal-extension/
- [ ] Test implemented payment add-on integrations

---

Time-depending, additional features will continue to try to get as close to _full_ integration parity as practical:

- [ ] Research existing _coupon_ add-ons or features for CF7, if any.
- [ ] Determine AffiliateWP pro add-on compatibility tasks, such as tiered rates.
- [ ] Determine feasability of using CF7 as an affiliate registration form alternative.
- [ ] Docs
- [ ] Teamwide testing
- [ ] Customer beta, if possible
- [ ] Release post written
- [ ] Launch
- [ ] Set affwp-docs channel topic to _Its been `int` days since the last CF7 ticket_

**Comment by pippinsplugins** [comment]
`Via the CF7 integration as well as standalone hooks such as wpcf7_contact_form_properties, define functionality consistent with the AFNF for Ninja Forms 2.x integration (inline meta boxes and options).`

That doesn't need to happen. For this issue we're only looking at referral tracking for Contact Form 7, not affiliate account registration.

**Comment by sumobi** [comment]
Related:

WP Affiliate

https://www.tipsandtricks-hq.com/wordpress-affiliate/how-to-capture-leads-using-the-wp-affiliate-platform-215

Affiliates Pro
https://wordpress.org/plugins/affiliates-contact-form-7/
http://www.itthinx.com/plugins/affiliates-pro-contact-form-7/

**Comment by ramiabraham** [comment]
(WIP https://github.com/AffiliateWP/AffiliateWP/pull/1848)

@AffiliateWP/core Should we support both of those PayPal add-ons? Coding one effectively codes the other with some misc minor changes needed.

**Comment by sumobi** [comment]
@ramiabraham If it's not too much trouble then yeah let's do it!

**Comment by pippinsplugins** [comment]
I agree, let's do both.

**Comment by ramiabraham** [comment]
Progress update!

Today I focused on getting a bit more granular with methods, to make later customization easier for customers, and increase the stability of the integration. Specifically:

- Focuses on keeping options and settings within AffWP core, instead of depending on settings pages within CF7 or its’ various add-ons.
- Uses native AffWP settings
- Defines numerous small, single-purpose methods to increase stability, such as defining supported gateway add-ons (ships with the two PayPal add-ons for now) separate from any _logic_ regarding those add-ons.
- Defines a number of filters for adjusting as needed, mostly in anticipation of customers using the numerous types of CF7 payment gateway add-ons available.
- Direct querying of CF7 forms via standard WP queries, instead of the various`WPCF7_` class methods.

---

Next steps are:

- Complete the re-written PayPal integrations (needs more checks to post meta and some related guard clauses).
- Move any validation checks away from the CF7 form object _itself_, in favor of just querying the _Flamingo_ post type object on creation via `wp_insert_post`.
- Test test test

**Comment by pippinsplugins** [comment]
Getting close :)

Ready to test whenever you are!

**Comment by ramiabraham** [comment]
Question (and update, particularly on what's taking so long):

Of the two notable PayPal add-ons for CF7, neither really _works_:

the first, https://wordpress.org/plugins/contact-form-7-paypal-add-on/, uses a static redirect which hooks on the `wpcf7_before_send_mail` action. This prevents _Flamingo_ from recording a submission, which it does via generating a post type `flamingo_inbound`.

What they need there is the _return to merchant_, along with related validation via nonce etc. The redirect just leaves you on PayPal - there's no information recorded on whether or not the transaction actually occurred.

---
 
The second CF7 PayPal add-on, https://wordpress.org/plugins/contact-form-7-paypal-extension/, is not compatible with CF7 version `4.6` or higher, and hasn't been updated in 5 months.

The issue is that `wpcf7_add_shortcode` was deprecated, and you now must use `wpcf7_add_form_tag` to add any functionality to the CF7 shortcode editor.

---

There are some options we have here:

- 1. Create our own standalone PayPal add-on for CF7 and give it away for free. Shouldn't take long, since one working day or less. there isn't actually anything to write form scratch here. Question is whether that support burden is worth taking on.

- 2. Inquire with the developers of both PayPal add-ons, and offer fixes etc, and move to other integrations for now.

- 3. Create workaround for this _within_ the AffiliateWP CF7 integration. That's what I've done today, but it's not pretty, so wanted to check before continuing.

**Comment by pippinsplugins** [comment]
I vote for 2 and 3. I'm not interested in maintaining our own PayPal add-on for such a large userbase.

**Comment by DrewAPicture** [comment]
I agree on a combination of 2 & 3. Our best bet is to stick a shim in AffWP and then contribute the fix(es) upstream to one or both of the established plugins.

**Comment by ramiabraham** [comment]
(I've reached out to both plugin authors; shims in progress for core).

**Comment by ramiabraham** [comment]
Mostly through this! Implementing some rewrites via a meeting yesterday. Will be working on it more tonight and later this weekend. (at https://github.com/AffiliateWP/AffiliateWP/pull/1848/commits/ef061bf66aaf793c637cfd402a8b840d8fb5e611 as of a few minutes ago).

**Comment by pippinsplugins** [comment]
@ramiabraham great!

**Comment by ramiabraham** [comment]
@AffiliateWP/core Update on CF7 finally -

- Flamingo is removed due to lack of hooks (for now). But Takayuki, the CF7 developer, is planning on adding some stuff to make this possible - mostly hooks, to inject meta into `flamingo_inbound` post types.

- The two [above-noted PayPal add-ons](https://github.com/AffiliateWP/AffiliateWP/issues/1336#issuecomment-268393865) have partial support; a fair bit of non-ideal, roundabout methods were required, but it's not _terrible_. 

- Takayuki is planning on creating an official PayPal CF7 add-on, which we'll of course support!

### Ready for (partial) testing:

#### PayPal add-on 1:

- PR https://github.com/AffiliateWP/AffiliateWP/pull/1848
- Install and activate [Contact Form 7](https://wordpress.org/plugins/contact-form-7/)
- Install and Activate [Contact Form 7 - Paypal Add-on](https://wordpress.org/plugins/contact-form-7-paypal-add-on)
- Activate the CF7 AffWP integration in _AffiliateWP &rarr; Settings &rarr; Integrations_.
- Configure PayPal settings in _Contact &rarr; PayPal Settings_.
    - Make sure to define a _success_ and _cancel_ url*.
- Create a CF7 form and configure the PayPal add-on settings in the PayPal tab.
- A _Contact Form 7_ config tab will appear in _AffiliateWP &rarr; Settings_. Select the form being used for referral testing, or _enable referrals for all Contact Form 7 forms_. Save.
- Test form submission; referrals will be generated using form object meta.

_Note: _safely_ injecting query args into the static redirect is still a WIP, so return and cancel referral updates are not yet working. The wip relies on `add_query_arg` - search for this in `../integrations/class-contactform7.php` for what I'm presently trying, and in `redirect.php` in the _Contact Form 7 - PayPal Add-on_ plugin.

**Comment by pippinsplugins** [comment]
@ramiabraham quick test all worked except for marking the referral as complete. Is that still WIP?

**Comment by ramiabraham** [comment]
@pippinsplugins Yes just waiting on CF7 reply/update to see what adjustments are required.

**Comment by pippinsplugins** [comment]
@ramiabraham where are we here? Did we get the less-than-optimal work around put back in?

**Comment by ramiabraham** [comment]
@pippinsplugins Yes - I'm presently adding this in and testing. As of a few minutes ago (https://github.com/AffiliateWP/AffiliateWP/pull/1848/commits/ab2e4c3fe122fc7ac3ca9daa9ee6778e436c3e4c), PayPal meta via the `contact-form-7-paypal-add-on` plugin is injected via js, using the ajax method `Affiliate_WP_Contact_Form_7::ajax_get_paypal_meta()`, via the `affwp_cf7_ajax()` wrapper.

Remaining is ensuring this data is parsed correctly in the cancel/return actions (in progress).

**Comment by ramiabraham** [comment]
@pippinsplugins Ready for review finally ([PR](https://github.com/AffiliateWP/AffiliateWP/pull/1848)). The only update/change to the above-defined integration usage instructions are that referrals as completed/rejected is now implemented.

**Comment by sumobi** [comment]
@ramiabraham I tested locally and can see the query string in the return URL but the referral stayed as pending. Does it only work on a live site?

When visiting the form on the front-end I'm seeing:

```
 PHP Notice:  Undefined variable: form_id in /wp-content/plugins/affiliatewp/includes/integrations/class-contactform7.php on line 776
```

Last line of stack trace:

`Affiliate_WP_Contact_Form_7->mark_referral_complete() /wp-includes/class-wp-hook.php:298`

When the form is submitted I see:

``` 
PHP Notice:  Undefined variable: current_page_id in /wp-content/plugins/affiliatewp/includes/integrations/class-contactform7.php on line 758
```

Last line of stack trace:

`Affiliate_WP_Contact_Form_7->add_pending_referral() /wordpress/wp-includes/class-wp-hook.php:298`

+ the first notice above repeated

**Comment by ramiabraham** [comment]
@AffiliateWP/core Ready for testing again.

**Comment by sumobi** [comment]
@ramiabraham Nice, all notices previously reported are gone. Once I arrived back to my thanks page the referral was updated to unpaid, so I'm going to assume the answer to my question above is "it will work locally" 👍 

URL was: `/thankyou/?form_id=6&sub_time=1484807544&amount=100&description=My%20Product&sku=1234referral_id%3d4`

However, the referral's amount was not updated and stays at $0.00:

![screen shot 2017-01-19 at 7 38 02 pm](https://cloud.githubusercontent.com/assets/52581/22096180/d673573c-de7e-11e6-9fb5-5ad1c1bf8c98.png)

One more thing, if the customer does not click to return to the site from PayPal, I'm guessing the referral will never be updated?

**Comment by ramiabraham** [comment]
@sumobi Yes,  testing should function on local development sites. My apologies for missing your prior question. Thanks for your patience!

Try encoding the `=` after `referral_id`? I'll check it out as well.

**Comment by ramiabraham** [comment]
@AffiliateWP/core Ready for testing yet again!

@sumobi Found the issue - I was missing an ampersand in the url string.

**Comment by pippinsplugins** [comment]
Worked great for me!

Just one thing I'd like see: update the referral on success to set the `reference` to the transaction ID from PayPal (it's in the URL).

**Comment by ramiabraham** [comment]
I don’t get the success transaction ID. What does yours look like?

http://affwp.dev/cf7/pp1-success/?form_id=1870&sub_time=1484837182&amount=3%2e00&description=Product%20description&sku=sku%2d12345&referral_id=46

**Comment by pippinsplugins** [comment]
Two quick issues:

1. Looks like the settings to enable referrals doesn't work 100%.

I don't have any forms enabled but a pending referral was still created:

![screen shot 2017-01-19 at 9 39 49 am](https://cloud.githubusercontent.com/assets/1034109/22113337/894274be-de2b-11e6-8de7-852b8132ea77.png)

2. Cancelled URL does not appear to rejecting the referrals any more.

**Comment by pippinsplugins** [comment]
Successful purchase also did not update to unpaid.

Regarding the transaction ID, my return URL was:

https://edd.com/cf7/purchase/success/?form_id=5&sub_time=1484840715&amount=1&description=Test&sku=&referral_id=6&amt=1.00&cc=USD&cm=6&item_name=Test&st=Completed&tx=4S132040PX028550Y

**Comment by pippinsplugins** [comment]
Ok this is ready for additional testing.

**Comment by ramiabraham** [comment]
Getting `array_key_exists() expects parameter 2 to be array, boolean given in .../affiliate-wp/includes/integrations/class-contactform7.php on line 216`.

Full trace:

`wpcf7_control_init() /app/public/wp-includes/class-wp-hook.php:298
[19-Jan-2017 17:10:21 UTC] PHP  10. wpcf7_submit_nonajax() /app/public/wp-content/plugins/contact-form-7/includes/controller.php:21
[19-Jan-2017 17:10:21 UTC] PHP  11. WPCF7_ContactForm->submit() /app/public/wp-content/plugins/contact-form-7/includes/controller.php:122
[19-Jan-2017 17:10:21 UTC] PHP  12. do_action() /app/public/wp-content/plugins/contact-form-7/includes/contact-form.php:676
[19-Jan-2017 17:10:21 UTC] PHP  13. WP_Hook->do_action() /app/public/wp-includes/plugin.php:453
[19-Jan-2017 17:10:21 UTC] PHP  14. WP_Hook->apply_filters() /app/public/wp-includes/class-wp-hook.php:323
[19-Jan-2017 17:10:21 UTC] PHP  15. Affiliate_WP_Contact_Form_7->add_pending_referral() /app/public/wp-includes/class-wp-hook.php:298
[19-Jan-2017 17:10:21 UTC] PHP  16. Affiliate_WP_Contact_Form_7->form_enabled() /app/public/wp-content/plugins/affiliate-wp/includes/integrations/class-contactform7.php:364
[19-Jan-2017 17:10:21 UTC] PHP  17. array_key_exists() /app/public/wp-content/plugins/affiliate-wp/includes/integrations/class-contactform7.php:216
[19-Jan-2017 17:10:21 UTC] PHP Warning:  array_key_exists() expects parameter 2 to be array, boolean given in /app/public/wp-content/plugins/affiliate-wp/includes/integrations/class-contactform7.php on line 216`

**Comment by ramiabraham** [comment]
@pippinsplugins Pushed a fix for the above; ready for testing.

**Comment by pippinsplugins** [comment]
@ramiabraham That commit won't fix the issue. `$this->get_enabled_forms()` needs to be adjusted to always return an array instead of boolean when empty.

**Comment by ramiabraham** [comment]
Sorry! Updated.

**Comment by pippinsplugins** [comment]
Merged into `release/2.0`!

**Comment by pippinsplugins** [comment]
Re-opening to support 0.00 referrals when PayPal is not enabled.

**Comment by pippinsplugins** [comment]
@sumobi please retest 0.00 and non 0.00 referrals on `release/2.0` please.
